## Table of contents

### 01 Importing libraries incl visualization libraries
### 02 Creating crosstabs
### 03 Task 4.10 part 1
- Task 4.10, question 2
- Task 4.10, question 3a
- Task 4.10, question 3b
- Task 4.10, question 4
- Exporting dataframes

### 01 Importing libraries incl visualization libraries

In [1]:
# import libraries incl visualization libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# import path:
path = r'C:\Users\svand\Documents\01-02-2024 Instacart Basket Analysis'

In [3]:
ords_prods_cust_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'customers_orders_products_merged.pkl'))

In [4]:
ords_prods_cust_merge.shape

(32404859, 32)

### 02 Creating crosstabs

In [5]:
# create a crosstab for excel
crosstab = pd.crosstab(ords_prods_cust_merge['days_since_prior_order'], ords_prods_cust_merge['order_number'], dropna = False)

In [6]:
crosstab.to_clipboard()

### 03 Task 4.10 part 1

In [8]:
ords_prods_cust_merge.head()

user_id first_name   surname  gender     state  age date_joined  children  \
0   26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01         3   
1   26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01         3   
2   26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01         3   
3   26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01         3   
4   26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01         3   

  family_status  annual_income  ...    price_range_loc    busiest_days  \
0       married         165665  ...  Mid-range product    Busiest days   
1       married         165665  ...  Mid-range product  Regularly busy   
2       married         165665  ...  Mid-range product    Busiest days   
3       married         165665  ...  Mid-range product  Regularly busy   
4       married         165665  ...  Mid-range product    Slowest days   

   busiest_hours  max_order  loyalty_flag  average_price_product  \
0    Most orders          8  New customer               7.988889   
1    Most orders          8  New customer               7.988889   
2    Most orders          8  New customer               7.988889   
3    Most orders          8  New customer               7.988889   
4    Most orders          8  New customer               7.988889   

   spending_flag  median_days_between_orders order_frequency_flag  _merge  
0    Low spender                        19.0     Regular customer    both  
1    Low spender                        19.0     Regular customer    both  
2    Low spender                        19.0     Regular customer    both  
3    Low spender                        19.0     Regular customer    both  
4    Low spender                        19.0     Regular customer    both  

[5 rows x 32 columns]

#### 2. Consider any security implications that might exist for this new data. You’ll need to address any PII data in the data before continuing your analysis.

In [9]:
ords_prods_cust_merge.columns

Index(['user_id', 'first_name', 'surname', 'gender', 'state', 'age',
       'date_joined', 'children', 'family_status', 'annual_income', 'order_id',
       'order_number', 'orders_day_of_week', 'order_hour_of_day',
       'days_since_prior_order', 'product_id', 'add_to_cart_order',
       'reordered', 'product_name', 'aisle_id', 'department_id', 'prices',
       'price_range_loc', 'busiest_days', 'busiest_hours', 'max_order',
       'loyalty_flag', 'average_price_product', 'spending_flag',
       'median_days_between_orders', 'order_frequency_flag', '_merge'],
      dtype='object')

##### Both first_name and surname are PII data, so those columns have to be removed.

In [7]:
df_instacard_merged = ords_prods_cust_merge.drop(columns = ['first_name', 'surname'])

In [8]:
# check if columns are dropped:
df_instacard_merged.head()

user_id  gender     state  age date_joined  children family_status  \
0   26711  Female  Missouri   48  2017-01-01         3       married   
1   26711  Female  Missouri   48  2017-01-01         3       married   
2   26711  Female  Missouri   48  2017-01-01         3       married   
3   26711  Female  Missouri   48  2017-01-01         3       married   
4   26711  Female  Missouri   48  2017-01-01         3       married   

   annual_income order_id  order_number  ...    price_range_loc  \
0         165665  2543867             5  ...  Mid-range product   
1         165665  1285508             7  ...  Mid-range product   
2         165665  2578584             8  ...  Mid-range product   
3         165665   423547             2  ...  Mid-range product   
4         165665  2524893             3  ...  Mid-range product   

     busiest_days  busiest_hours  max_order  loyalty_flag  \
0    Busiest days    Most orders          8  New customer   
1  Regularly busy    Most orders          8  New customer   
2    Busiest days    Most orders          8  New customer   
3  Regularly busy    Most orders          8  New customer   
4    Slowest days    Most orders          8  New customer   

   average_price_product spending_flag  median_days_between_orders  \
0               7.988889   Low spender                        19.0   
1               7.988889   Low spender                        19.0   
2               7.988889   Low spender                        19.0   
3               7.988889   Low spender                        19.0   
4               7.988889   Low spender                        19.0   

   order_frequency_flag  _merge  
0      Regular customer    both  
1      Regular customer    both  
2      Regular customer    both  
3      Regular customer    both  
4      Regular customer    both  

[5 rows x 30 columns]

####  3a. Create a regional segmentation of the data. 

In [12]:
# check state details:
df_instacard_merged['state'].value_counts(dropna = False)

state
Pennsylvania            667082
California              659783
Rhode Island            656913
Georgia                 656389
New Mexico              654494
Arizona                 653964
North Carolina          651900
Oklahoma                651739
Alaska                  648495
Minnesota               647825
Massachusetts           646358
Wyoming                 644255
Virginia                641421
Missouri                640732
Texas                   640394
Colorado                639280
Maine                   638583
North Dakota            638491
Alabama                 638003
Kansas                  637538
Louisiana               637482
Delaware                637024
South Carolina          636754
Oregon                  636425
Arkansas                636144
Nevada                  636139
New York                635983
Montana                 635265
South Dakota            633772
Illinois                633024
Hawaii                  632901
Washington              632852
Mi

In [9]:
#Setting regions with the loc() function:
df_instacard_merged.loc[df_instacard_merged['state'].isin(['Maine', 'New Hampshire', 'Vermont', 'Massachusetts', 'Rhode Island', 'Connecticut', 'New York', 'Pennsylvania', 'New Jersey']),'region_loc'] = 'Northeast'

In [10]:
df_instacard_merged.loc[df_instacard_merged['state'].isin(['Wisconsin', 'Michigan', 'Illinois', 'Indiana', 'Ohio', 'North Dakota', 'South Dakota', 'Nebraska', 'Kansas', 'Minnesota', 'Iowa', 'Missouri']),'region_loc'] = 'Midwest'

In [11]:
df_instacard_merged.loc[df_instacard_merged['state'].isin(['Delaware', 'Maryland', 'District of Columbia', 'Virginia', 'West Virginia', 'North Carolina', 'South Carolina', 'Georgia', 'Florida', 'Kentucky', 'Tennessee', 'Mississippi', 'Alabama', 'Oklahoma', 'Texas', 'Arkansas', 'Louisiana']),'region_loc'] = 'South'

In [12]:
df_instacard_merged.loc[df_instacard_merged['state'].isin(['Idaho', 'Montana','Wyoming','Nevada','Utah','Colorado','Arizona','New Mexico','Alaska','Washington','Oregon','California','Hawaii']),'region_loc'] = 'West'

In [13]:
# checking the created 'region_loc' column:
df_instacard_merged['region_loc'].value_counts(dropna = False)

region_loc
South        10791885
West          8292913
Midwest       7597325
Northeast     5722736
Name: count, dtype: int64

In [18]:
df_instacard_merged.head()

user_id  gender     state  age date_joined  children family_status  \
0   26711  Female  Missouri   48  2017-01-01         3       married   
1   26711  Female  Missouri   48  2017-01-01         3       married   
2   26711  Female  Missouri   48  2017-01-01         3       married   
3   26711  Female  Missouri   48  2017-01-01         3       married   
4   26711  Female  Missouri   48  2017-01-01         3       married   

   annual_income order_id  order_number  ...    busiest_days  busiest_hours  \
0         165665  2543867             5  ...    Busiest days    Most orders   
1         165665  1285508             7  ...  Regularly busy    Most orders   
2         165665  2578584             8  ...    Busiest days    Most orders   
3         165665   423547             2  ...  Regularly busy    Most orders   
4         165665  2524893             3  ...    Slowest days    Most orders   

   max_order  loyalty_flag  average_price_product  spending_flag  \
0          8  New customer               7.988889    Low spender   
1          8  New customer               7.988889    Low spender   
2          8  New customer               7.988889    Low spender   
3          8  New customer               7.988889    Low spender   
4          8  New customer               7.988889    Low spender   

  median_days_between_orders  order_frequency_flag  _merge  region_loc  
0                       19.0      Regular customer    both     Midwest  
1                       19.0      Regular customer    both     Midwest  
2                       19.0      Regular customer    both     Midwest  
3                       19.0      Regular customer    both     Midwest  
4                       19.0      Regular customer    both     Midwest  

[5 rows x 31 columns]

#### 3b. Determine whether there’s a difference in spending habits between the different U.S. regions.

In [14]:
# Create a crosstab:
crosstab_region_spending_habits = pd.crosstab(df_instacard_merged['spending_flag'], df_instacard_merged['region_loc'], dropna = False)

In [15]:
crosstab_region_spending_habits

region_loc     Midwest  Northeast     South     West
spending_flag                                       
High spender    155975     108225    209691   160354
Low spender    7441350    5614511  10582194  8132559

##### Most high spending customers are in the South. Followed by West, Midwest and in the Northeast are the lowest number of high spenders
##### In the low spending category the numbers are in the same descending order, as wel as the number of customers in each region:
##### Number of customers in each region:
- South        10791885,
- West          8292913,
- Midwest       7597325,
- Northeast     5722736

##### Comparing those numbers, I can see there is nog big difference in spending habits.

##### 4. Create an exclusion flag for low-activity customers (customers with less than 5 orders) and exclude them from the data.

In [16]:
#Creating a exclusion flag for 'max_order' columnn
df_instacard_merged.loc[df_instacard_merged['max_order'] >= 5, 'exclusion_flag'] = 'normal-activity customers'

In [17]:
df_instacard_merged.loc[df_instacard_merged['max_order'] < 5, 'exclusion_flag'] = 'low-activity customers'

In [18]:
#checking the results 
df_instacard_merged[['max_order', 'exclusion_flag']].head(60)

max_order             exclusion_flag
0           8  normal-activity customers
1           8  normal-activity customers
2           8  normal-activity customers
3           8  normal-activity customers
4           8  normal-activity customers
5           8  normal-activity customers
6           8  normal-activity customers
7           8  normal-activity customers
8           8  normal-activity customers
9           8  normal-activity customers
10          8  normal-activity customers
11          8  normal-activity customers
12          8  normal-activity customers
13          8  normal-activity customers
14          8  normal-activity customers
15          8  normal-activity customers
16          8  normal-activity customers
17          8  normal-activity customers
18         12  normal-activity customers
19         12  normal-activity customers
20         12  normal-activity customers
21         12  normal-activity customers
22         12  normal-activity customers
23         12  normal-activity customers
24         12  normal-activity customers
25         12  normal-activity customers
26         12  normal-activity customers
27         12  normal-activity customers
28         12  normal-activity customers
29         12  normal-activity customers
30         12  normal-activity customers
31         12  normal-activity customers
32         12  normal-activity customers
33         12  normal-activity customers
34         12  normal-activity customers
35         12  normal-activity customers
36         12  normal-activity customers
37         12  normal-activity customers
38         12  normal-activity customers
39         12  normal-activity customers
40         12  normal-activity customers
41         12  normal-activity customers
42         12  normal-activity customers
43         12  normal-activity customers
44         12  normal-activity customers
45         12  normal-activity customers
46         12  normal-activity customers
47         12  normal-activity customers
48         12  normal-activity customers
49         12  normal-activity customers
50         12  normal-activity customers
51         12  normal-activity customers
52         12  normal-activity customers
53         12  normal-activity customers
54         12  normal-activity customers
55         12  normal-activity customers
56         12  normal-activity customers
57         12  normal-activity customers
58         12  normal-activity customers
59         12  normal-activity customers

In [19]:
df_instacard_merged[['max_order', 'exclusion_flag']].tail(120)

max_order          exclusion_flag
32404739          4  low-activity customers
32404740          4  low-activity customers
32404741          4  low-activity customers
32404742          4  low-activity customers
32404743          4  low-activity customers
...             ...                     ...
32404854          4  low-activity customers
32404855          4  low-activity customers
32404856          4  low-activity customers
32404857          4  low-activity customers
32404858          4  low-activity customers

[120 rows x 2 columns]

In [26]:
# Creating a sample for low-activity customers
df_low_act_customers = df_instacard_merged[df_instacard_merged['exclusion_flag'] =='low-activity customers']

In [27]:
# Creating a sample for normal-activity customers
df_nor_act_customers = df_instacard_merged[df_instacard_merged['exclusion_flag'] =='normal-activity customers']

In [23]:
df_instacard_merged.info

<bound method DataFrame.info of          user_id  gender     state  age date_joined  children family_status  \
0          26711  Female  Missouri   48  2017-01-01         3       married   
1          26711  Female  Missouri   48  2017-01-01         3       married   
2          26711  Female  Missouri   48  2017-01-01         3       married   
3          26711  Female  Missouri   48  2017-01-01         3       married   
4          26711  Female  Missouri   48  2017-01-01         3       married   
...          ...     ...       ...  ...         ...       ...           ...   
32404854   80148  Female  New York   55  2020-04-01         1       married   
32404855   80148  Female  New York   55  2020-04-01         1       married   
32404856   80148  Female  New York   55  2020-04-01         1       married   
32404857   80148  Female  New York   55  2020-04-01         1       married   
32404858   80148  Female  New York   55  2020-04-01         1       married   

          annual_in

In [28]:
len(df_low_act_customers) + len(df_nor_act_customers)

32404859

### 04 Exporting dataframes

In [29]:
df_low_act_customers.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'low_activity_customers.pkl'))

In [30]:
df_nor_act_customers.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'normal_activity_customers.pkl'))